engine-nextgen/src/main/kotlin/graphql/nadel/validation/NadelHydrationValidation.kt

```diff
@@ -91,40 +74,44 @@ internal class NadelHydrationValidation(
     private fun getOutputTypeIssues(
         parent: NadelServiceSchemaElement,
         overallField: GraphQLFieldDefinition,
-        actorService: Service,
         actorField: GraphQLFieldDefinition,
-        isPolymorphicHydration: Boolean,
+        hasMoreThanOneHydration: Boolean,
     ): List<NadelSchemaValidationError> {
         // Ensures that the underlying type of the actor field matches with the expected overall output type
-        var overallType = overallField.type.unwrapAll()
-        if (isPolymorphicHydration) {
-            if (overallType is GraphQLUnionType) {
-                val possibleObjectTypes = overallType.types
-                val actorFieldReturnType = actorField.type.unwrapAll().name
-                val overallTypeMatchingActorFieldReturnType = possibleObjectTypes.find {
-                    actorFieldReturnType == getUnderlyingType(it, actorService)?.name
+        val overallType = overallField.type.unwrapAll()
+
+        // Polymorphic hydration must have union output
+        if (hasMoreThanOneHydration && overallType !is GraphQLUnionType) {
+            return listOf(FieldWithPolymorphicHydrationMustReturnAUnion(parent, overallField))
+        }
+
+        val acceptableOutputTypes: List<GraphQLNamedOutputType> = when (overallType) {
+            is GraphQLUnionType -> overallType.types + overallType
+            is GraphQLInterfaceType -> overallSchema.getImplementations(overallType) + overallType
+            else -> listOf(overallType as GraphQLNamedOutputType)
+        }
+
+        val actorOutputTypes: List<GraphQLNamedOutputType> = when (val actorOutputType = actorField.type.unwrapAll()) {
+            is GraphQLUnionType -> actorOutputType.types
+            is GraphQLInterfaceType -> overallSchema.getImplementations(actorOutputType)
+            else -> listOf(actorOutputType as GraphQLNamedOutputType)
+        }
+
+        val typeValidation = actorOutputTypes
+            // Find incompatible output types
+            .filter { actorOutputType ->
+                acceptableOutputTypes.none { acceptableOutputType ->
+                    typeValidation.isAssignableTo(lhs = acceptableOutputType, rhs = actorOutputType)
```

Comment:
Made some changes to the way the output type is validated. Basically we can validate stuff like

```graphql
type Query {
  userById(id: ID!): User
}
type Issue {
  assignee: AbstractUser @hydrated(field: "userById")
}
interface AbstractUser {
}
type User implements AbstractUser {
}
```

i.e. type doesn't exactly match up, but we do a smarter check on whether they're compatible.